docs/polymorphic-table-types.md

```diff
@@ -0,0 +1,198 @@
+
+3. The implementation is opaque and can be modified later without worry.
+4. Mutations require yet another function, a setter, with the cons listed before.
+5. Iterations requires yet another function, or at least stored iterator, reinforcing the cons listed before.
+
+In short, tables are mutable and iterable, functions are not.
+
+At this point if one is not satisfied with the extra baggage that goes behind a type-safe access or write of this nature, they will simply lie to the type checker or compromise and move on. This is exactly what we wish to avoid.
+
+This proposal targets users already neck-deep in the type-system supporting libraries and tools, unable to achieve safety without compromising performance or their api and affecting thousands of users, or the occasional type-savvy user getting that perfect type for their data structure and feeling accomplished.
+
+### Specific Indexer Pattern
+
+We'll demonstrate getting components from an entity using a minified ECS api from a library currently in production. This library depends on accessing components from tables using their constructors as indexers. The constructors contain the type information of the component, so it is theoretically possible to infer the type of the resulting component from the constructor, but this is not possible as of now because polymorphic tables don't exist.
+
+```lua
+-- <C>(() -> C) -> Factory<C>
+world.factory
+
+-- <'Foo'>(() -> 'Foo') -> Factory<'Foo'> ==> Factory<'Foo'>
+local A = world.factory(function()
+	return 'Foo'
+end)
+
+-- <'Bar'>(() -> 'Bar') -> Factory<'Bar'> ==> Factory<'Bar'>
+local B = world.factory(function()
+	return 'Bar'
+end)
+
+-- number
+local e = world.entity()
+
+-- 'Foo'
+A.add(e)
+
+-- 'Bar'
+B.add(e)
+
+-- { [Factory<any>]: any? } 	(no possible way to extract the generic)
+local components = world.get(e)
+
+-- { [Factory<any>]: any? }[Factory<'Foo'>] ==> any?
+local a = components[A]
+
+-- { [Factory<any>]: any? }[Factory<'Bar'>] ==> any?
+local b = components[B]
+```
+
+We've lost our type information of `a` and `b` in the last table step!
+
+-   "Why not use a getter?"
+    -   Mentioned before, API bloat or performance concerns, getting components from this table usually happens during iterations and thousands of times a frame.
+-   "Why not implement better metamethod inference?"
+    -   The `components` table guarantees users may define their own metatables. Metatables are not an elegant, reliable, or wholly-acceptable solution to this lack of types. However better inference is always a good thing! This is a great idea for a _different_ RFC.
+
+## Design
+
+We propose naturally extending this type-mapping capability to tables in the backwards-compatible form `<T>{ [F<T>]: G<T> }`. We recycle our familiar syntax of polymorphic functions and use them in a very similar fashion. If one understands polymorphic functions, they will understand this as well. It is backwards compatible as it is currently invalid syntax.
```

Comment:
I'm unsure of how iterations over this kind of table can be inferred at all. Given
```lua
type A = <T>{ [F<T>]: G<T> }
local a: A = nil :: any -- lie to the type system to initialize this without errors

for k, v in a do
    -- what are k and v?
end
```
What does this do? Since `T` is unbounded (like all generics), that means its upper bound is `unknown` - but that doesn't necessarily mean that $\text{F}\verb|<|\text{T}\verb|>| <:\text{unknown}$, and the same for `G<T>`. Perhaps `F<T>` is defined as just `any`, or perhaps it's a type family (which is permitted to expand to literally anything). At the same time, we can't simply _not substitute_ `T` - a bare generic type is incompatible with basically everything except itself or other generics (in specific circumstances) - so I don't see a pair of types we can safely provide for `k` and `v` here.